src/RealDot.jl

```diff
@@ -1,5 +1,22 @@
 module RealDot
 
-# Write your package code here.
+using LinearAlgebra: LinearAlgebra
+
+export realdot
+
+"""
+    realdot(x, y)
+
+Compute `real(dot(x, y))` while avoiding computing the imaginary part if possible.
+
+This function can be useful if you work with derivatives of functions on complex
+numbers. In particular, this computation shows up in pullbacks for non-holomorphic
+functions.
+"""
+@inline realdot(x, y) = real(LinearAlgebra.dot(x, y))
+@inline realdot(x::Number, y::Number) = muladd(real(x), real(y), imag(x) * imag(y))
```

Comment:
On second thought, this will break for some hypercomplex number implementations, where whether or not to overload `imag` is less clear (and `imag` may return a vector or something else). e.g. Quaternions that defines its own `imag` function. Best to restrict this to `Complex`:
```suggestion
@inline realdot(x::Complex, y::Complex) = muladd(real(x), real(y), imag(x) * imag(y))
```